File: crates/sim/examples/qualify-tree.rs
//! Persistent collisionless population: the solver cannot drop below the tree threshold.
//! Optional arguments `tree` and/or `exact` run only those solvers, so CI can
//! run each on its own machine; `scripts/qualify-tree.mjs` merges the outputs.
//! With neither, both run in turn.
//!
//! Each run also records every body's first close encounter: a separation
//! below `ENCOUNTER_HILL` mutual Hill radii. Such passes are physically
//! chaotic, so the gate judges tree accuracy on the bodies that have not had
//! one (and bounds how many have).
use celestial_sim::{benchmark::OrbitProbe, *};
use serde_json::json;
fn balances(s: &[f64]) -> [f64; 4] {
    let mut out = [0.; 4];
    for (i, a) in s.chunks_exact(5).enumerate() {
        out[0] += 0.5 * a[4] * (a[2] * a[2] + a[3] * a[3]);
        out[1] += a[4] * (a[0] * a[3] - a[1] * a[2]);
        out[2] += a[4] * a[2];
        out[3] += a[4] * a[3];
        for b in s.chunks_exact(5).skip(i + 1) {
            out[0] -=
                G * a[4] * b[4] / ((a[0] - b[0]).powi(2) + (a[1] - b[1]).powi(2) + 1e-8).sqrt();
        }
    }
    out
}
/// Close-encounter separation in mutual Hill radii: entering each other's
/// Hill sphere.
const ENCOUNTER_HILL: f64 = 1.0;
/// Marks, for every body without one yet, a separation below its encounter
/// distance at this tick: an x-sorted sweep, since the distances are small.
fn mark_encounters(s: &[f64], tick: u64, first: &mut [Option<u64>]) {
    let n = s.len() / 5;
    let (x0, y0, m0) = (s[0], s[1], s[4]);
    let r = |i: usize| (s[5 * i] - x0).hypot(s[5 * i + 1] - y0);
    let hill = |i: usize, j: usize| {
        ENCOUNTER_HILL * 0.5 * (r(i) + r(j)) * ((s[5 * i + 4] + s[5 * j + 4]) / (3.0 * m0)).cbrt()
    };
    let mut order: Vec<usize> = (1..n).collect();
    order.sort_unstable_by(|&a, &b| s[5 * a].total_cmp(&s[5 * b]));
    let (r_max, m_max) = (1..n).fold((0.0f64, 0.0f64), |(a, b), i| {
        (a.max(r(i)), b.max(s[5 * i + 4]))
    });
    let window = ENCOUNTER_HILL * r_max * (2.0 * m_max / (3.0 * m0)).cbrt();
    for (k, &i) in order.iter().enumerate() {
        for &j in &order[k + 1..] {
            if s[5 * j] - s[5 * i] > window {
                break;
            }
            let d = (s[5 * j] - s[5 * i]).hypot(s[5 * j + 1] - s[5 * i + 1]);
            if d < hill(i, j) {
                for b in [i, j] {
                    first[b].get_or_insert(tick);
                }
            }
        }
    }
}
fn main() {
    let mut w = World::new(Config {
        mission: None,
        ..Default::default()
    })
    .unwrap();
    w.apply(Command::SeedSwarm {
        count: 511,
        disorder: 0.,
    })
    .unwrap();
    // Cold low-mass disk isolates accumulation of tree error from unresolved hard encounters.
    let initial: Vec<f64> = w
        .bodies
        .iter()
        .flat_map(|b| {
            [
                b.pos.x,
                b.pos.y,
                b.vel.x,
                b.vel.y,
                if b.id == 0 { b.mass } else { b.mass * 0.001 },
            ]
        })
        .collect();
    let selected: Vec<String> = std::env::args().skip(1).collect();
    let solvers: Vec<bool> = [false, true]
        .into_iter()
        .filter(|&exact| {
            let name = if exact { "exact" } else { "tree" };
            selected.is_empty() || selected.iter().any(|s| s == name)
        })
        .collect();
    assert!(!solvers.is_empty(), "unknown solver: {selected:?}");
    // One solver per process in CI and in the release script, which run them
    // concurrently; WebAssembly under WASI has no threads.
    let runs: Vec<_> = solvers
        .iter()
        .map(|&exact| {
            let mut p = OrbitProbe::new(&initial, exact).unwrap();
            let mut samples = vec![];
            let mut first = vec![None; initial.len() / 5];
            for year in 1..=600u64 {
                // One tick at a time is bit-identical to `advance(512)`.
                for tick in 0..512 {
                    p.advance(1);
                    mark_encounters(&p.state(), (year - 1) * 512 + tick + 1, &mut first);
                }
                if year % 25 == 0 {
                    let s = p.state();
                    samples.push(json!({"year":year,"balances":balances(&s),"state":s}));
                    eprintln!("persistent 512, exact={exact}: {year} years");
                }
            }
            json!({"exact":exact,"samples":samples,"first_encounter_tick":first})
        })
        .collect();
    println!(
        "{}",
        json!({"schema":1,"physics":checkpoint::PHYSICS_ID,"bodies":512,"debris_earth_masses":0.016,"initial_balances":balances(&initial),"runs":runs})
    );
}

File: scripts/qualify-tree.mjs
import { writeFile } from 'node:fs/promises';
import assert from 'node:assert/strict';
import { readRaw } from './merge-raw.mjs';
// One raw file with both solvers, or one per solver from separate machines;
// the tree run comes first and the exact reference second.
const raw = await readRaw(process.argv.slice(2), (run) => String(run.exact));
raw.runs.sort((a, b) => Number(a.exact) - Number(b.exact));
assert.deepEqual(
  raw.runs.map((run) => run.exact),
  [false, true],
  'Both the tree and the exact reference must be qualified',
);
const initial = raw.initial_balances,
  failures = [];
const rows = raw.runs.map((run) => ({
  exact: run.exact,
  years: run.samples.at(-1).year,
  maxRelativeEnergy: Math.max(
    ...run.samples.map((s) => Math.abs((s.balances[0] - initial[0]) / initial[0])),
  ),
  maxAngularResidual: Math.max(...run.samples.map((s) => Math.abs(s.balances[1] - initial[1]))),
  maxMomentumResidual: Math.max(
    ...run.samples.map((s) => Math.hypot(s.balances[2] - initial[2], s.balances[3] - initial[3])),
  ),
}));
// Close encounters are physically chaotic: they magnify any difference
// between the runs, whatever its source. Tree accuracy is judged on the
// bodies that have not had one in either run, and few may have had one.
const TICKS_PER_YEAR = 512,
  MAX_ENCOUNTERED = 0.1,
  MAX_RMS_AU = 1e-4;
const firstEncounter = raw.runs[0].first_encounter_tick.map((tick, i) =>
  Math.min(tick ?? Infinity, raw.runs[1].first_encounter_tick[i] ?? Infinity),
);
const trajectory = raw.runs[0].samples.map((sample, index) => {
  const reference = raw.runs[1].samples[index];
  let all = 0,
    calm = 0,
    count = 0;
  for (let b = 1; b < raw.bodies; b++) {
    const i = 5 * b;
    const d2 =
      (sample.state[i] - reference.state[i]) ** 2 +
      (sample.state[i + 1] - reference.state[i + 1]) ** 2;
    all += d2;
    if (firstEncounter[b] > sample.year * TICKS_PER_YEAR) {
      calm += d2;
      count++;
    }
  }
  return {
    year: sample.year,
    rmsPositionAu: Math.sqrt(calm / count),
    encounteredBodies: raw.bodies - 1 - count,
    rmsAllBodiesAu: Math.sqrt(all / (raw.bodies - 1)),
  };
});
const encountered = firstEncounter
  .slice(1)
  .flatMap((tick, i) => (tick === Infinity ? [] : [{ body: i + 1, year: tick / TICKS_PER_YEAR }]));
for (const row of rows)
  if (
    row.years !== 600 ||
    row.maxRelativeEnergy > 2e-5 ||
    row.maxAngularResidual > 1e-9 ||
    row.maxMomentumResidual > 1e-9
  )
    failures.push(`Persistent tree balance gate failed: exact=${row.exact}`);
if (trajectory.some((s) => s.rmsPositionAu > MAX_RMS_AU))
  failures.push(`Tree/direct trajectory RMS exceeds ${MAX_RMS_AU} AU`);
if (encountered.length > MAX_ENCOUNTERED * (raw.bodies - 1))
  failures.push(`${encountered.length} bodies had close encounters; at most 10% may`);
const report = {
  schema: 1,
  physics: raw.physics,
  bodies: raw.bodies,
  debrisEarthMasses: raw.debris_earth_masses,
  scope:
    'Constant 512-body collisionless cold disk; full 600-year tree lifetime, independently summed direct-force comparison at the same timestep. Low debris mass isolates force accumulation from hard encounters. Trajectory RMS covers bodies without a close encounter (inside one mutual Hill radius) in either run so far; at most 10% of bodies may have one.',
  rows,
  encountered,
  trajectory,
  passed: !failures.length,
  failures,
};
await writeFile('tree-qualification.json', JSON.stringify(report, null, 2) + '\n');
console.log(JSON.stringify(report, null, 2));
if (failures.length) process.exitCode = 1;
